### PR TITLE
[fuzz] add fuzz coverage CI check

### PR DESCRIPTION
**Pull request by asraa**
Commit Message: Add `bazel.fuzz_coverage` to CI. Publishes fuzz coverage report in `coverage_publish`
Risk Level: Medium
Testing: Local testing of `./ci/run_envoy_docker.sh './ci/do_ci.sh bazel.fuzz_coverage'`, as well as `FUZZ_COVERAGE=true test/run_envoy_bazel_coverage.sh` with single test args/directory, VALIDATE on/off.
Fixes: https://github.com/envoyproxy/envoy/issues/9573
Docs: Updated bazel/README.md and fuzz/README.md for docs about running local coverage.

Signed-off-by: Asra Ali <asraa@google.com>

### CODE REVIEW

**Comment by asraa** [comment]
@htuch @lizan This is my first stab at adding to CI. `coverage_publish` publishes both reports, and fuzz coverage writes out to `/generated/fuzz_coverage` now. I made the fuzz coverage at 27% right now, hoping to increase this as we see gaps/add fuzz targets over Q2/Q3.

**Comment by htuch** [comment]
@asraa Very cool. LMK when this passes; I'd like to take a look at the artifacts in CI. Are you planning on adding a gating check on overall fuzz coverage to encourage folks to write more fuzzers in this PR?

**Comment by asraa** [comment]
CI failure is interesting. An easy solution is to run the gtest mode of the fuzz target when running `bazel coverage`. 

I pushed the server corpus entry that triggers the `ApiBoostRetryException` and failure to the PR. It crashes with an uncaught exception. Reproduce with:
`bazel run --config=libc++  --test_output=all --test_arg=/usr/local/google/home/asraa/git/envoy2/test/server/server_corpus --test_arg=-runs=0  //test/server/config_validation:config_fuzz_test_with_libfuzzer`
This runs the test against the seed corpus with the fuzzing driver.

* It is important to run the target `_with_libfuzzer` that links the fuzzing main. It does not repro with using standard gtest main. 
* It must use `--config=libc++`. Without it, it succeeds. 
* If you add `--config clang-asan` in addition to `--config=libc++`, it will succeed.

@yanavlasov Does this indicate that this might be a compiler bug?
stack trace shows the exception is caught in `tryWithApiBoosting` executing the lambda [here](https://github.com/envoyproxy/envoy/blob/49efb9841a58ebdc43a666f55c445911c8e4181c/source/common/protobuf/utility.cc#L154):
```
terminating with uncaught exception of type Envoy::(anonymous namespace)::ApiBoostRetryException
==36646== ERROR: libFuzzer: deadly signal
    #0 0x2e75450 in __sanitizer_print_stack_trace (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x2e75450)
    #1 0x2e20818 in fuzzer::PrintStackTrace() (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x2e20818)
    #2 0x2e05f83 in fuzzer::Fuzzer::CrashCallback() (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x2e05f83)
    #3 0x7f8b9648f38f  (/lib/x86_64-linux-gnu/libpthread.so.0+0x1138f)
    #4 0x7f8b960e9427 in raise (/lib/x86_64-linux-gnu/libc.so.6+0x35427)
    #5 0x7f8b960eb029 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x37029)
    #6 0x8aafb0a in abort_message (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x8aafb0a)
    #7 0x8aafa35 in demangling_terminate_handler() (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x8aafa35)
    #8 0x8aaf892 in std::__terminate(void (*)()) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x8aaf892)
    #9 0x8aaeee5 in __cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x8aaeee5)
    #10 0x8aaee7e in __cxa_throw (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x8aaee7e)
    #11 0x72d19a6 in Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1::operator()(google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion) const (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72d19a6)
    #12 0x72e25a9 in decltype(std::__1::forward<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1&>(fp)(std::__1::forward<google::protobuf::Message&>(fp0), std::__1::forward<Envoy::(anonymous namespace)::MessageVersion>(fp0))) std::__1::__invoke<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1&, google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion>(Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1&, google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion&&) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72e25a9)
    #13 0x72e24fa in void std::__1::__invoke_void_return_wrapper<void>::__call<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1&, google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion>(Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1&, google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion&&) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72e24fa)
    #14 0x72e248a in std::__1::__function::__alloc_func<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1, std::__1::allocator<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1>, void (google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion)>::operator()(google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion&&) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72e248a)
    #15 0x72e0e6c in std::__1::__function::__func<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1, std::__1::allocator<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1>, void (google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion)>::operator()(google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion&&) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72e0e6c)
    #16 0x72d4d02 in std::__1::__function::__value_func<void (google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion)>::operator()(google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion&&) const (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72d4d02)
    #17 0x72d4c67 in std::__1::function<void (google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion)>::operator()(google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion) const (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72d4c67)
    #18 0x72d00bb in Envoy::(anonymous namespace)::tryWithApiBoosting(std::__1::function<void (google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion)>, google::protobuf::Message&) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72d00bb)
```

**Comment by lizan** [comment]
yeah I think we might want to hold this until we sort out multi-binary coverage #10909, I think that will get rid of most issues we hit with coverage.

**Comment by asraa** [comment]
marking as draft, for both the clang issue and multibinary coverage

**Comment by asraa** [comment]
This is ready for review!

**Comment by asraa** [comment]
:/ There's a regression from a recent PR submitted. Will ping when it's resolved.

**Comment by asraa** [comment]
H/2 test timeout is blocking CI: https://github.com/envoyproxy/envoy/issues/11290. Will merge master when fix is merged

**Comment by lizan** [comment]
@asraa PTAL at https://github.com/envoyproxy/envoy/pull/11251 as well, I think we are ready to use fuzz in azure pipelines instead of circle.

**Comment by asraa** [comment]
Merged, move the fuzz coverage job to azp

**Comment by asraa** [comment]
Will debug. Got 
```
sent 58,277,684 bytes  received 86,579 bytes  116,728,526.00 bytes/sec
total size is 57,915,954  speedup is 0.99
Code coverage 29.1 is good and higher than limit of 27.0
HTML coverage report is in /source/generated/fuzz_coverage/index.html
$TEST_TMPDIR defined: output root default is '/build/tmp' and max_idle_secs default is '15'.
WARNING: The following rc files are no longer being read, please transfer their contents or import their path into one of the standard rc files:
/source/tools/bazel.rc
cp: cannot stat '/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/command.profile': No such file or directory
```
at the end of the fuzz coverage script, but when publishing, I got `##[error]The operation was canceled.`. 

Slightly stumped, but will check

**Comment by asraa** [comment]
PTAL! I think this ready for review

File: .azure-pipelines/pipelines.yml [hunk 1]
```diff
@@ -103,7 +103,28 @@ jobs:
...
       - script: ci/run_envoy_docker.sh 'ci/upload_gcs_artifact.sh /build/envoy/generated/coverage coverage'
-        displayName: "Upload Report to GCS"
+        displayName: "Upload Coverage Report to GCS"
+        env:
+          ENVOY_DOCKER_BUILD_DIR: $(Build.StagingDirectory)
+          GCP_SERVICE_ACCOUNT_KEY: $(GcpServiceAccountKey)
+          GCS_ARTIFACT_BUCKET: $(GcsArtifactBucket)
+        condition: always()
+
+  - job: fuzz_coverage
```
Comment (lizan) [inline .azure-pipelines/pipelines.yml:113]: I think you can make this [coverage, fuzz_coverage] to matirx? (see bazel job above)

File: .azure-pipelines/pipelines.yml [hunk 2]
```diff
@@ -94,16 +94,31 @@ jobs:
     dependsOn: ["format"]
     timeoutInMinutes: 360
     pool: "x64-large"
+    strategy:
+      maxParallel: 2
+      matrix:
+        coverage:
+          CI_TARGET: "bazel.coverage"
```
Comment (lizan) [inline .azure-pipelines/pipelines.yml:101]: make this CI_TARGET: "coverage"/"fuzz_coverage" then you can use it in upload step as well?

Comment (lizan) [reply .azure-pipelines/pipelines.yml:101]: Can you make this as displayName: "Linux-x64" in line 93? then the GH check name should match existing.

File: bazel/README.md [hunk 3]
```diff
@@ -571,21 +571,31 @@ test/run_envoy_bazel_coverage.sh
...
+This generates a coverage report for fuzz targets after running the target for one minute against fuzzing engine libfuzzer using its coprus as initial seed inputs. The full coverage report will be available in `generated/fuzz_coverage/coverage.html`.
+
 Coverage for every PR is available in Circle in the "artifacts" tab of the coverage job. You will
 need to navigate down and open "coverage.html" but then you can navigate per normal. NOTE: We
 have seen some issues with seeing the artifacts tab. If you can't see it, log out of Circle, and
 then log back in and it should start working.
 
 The latest coverage report for master is available
-[here](https://storage.googleapis.com/envoy-coverage/report-master/index.html).
+[here](https://storage.googleapis.com/envoy-coverage/report-master/index.html). The latest fuzz coverage report for master is available [here](https://storage.googleapis.com/envoy-fuzz-coverage/report-master/index.html)
```
Comment (lizan) [inline bazel/README.md:586]: merge master, this should be in https://storage.googleapis.com/envoy-postsubmit/master/fuzz_coverage/index.html

**Comment by lizan** [comment]
LGTM modulo fuzz_coverage CI is failing.

/wait

**Comment by asraa** [comment]
Looks like I hit an ASAN abort, not sure what changed
```
2020-06-01T19:24:21.2160138Z + test/server/config_validation/config_fuzz_test_with_libfuzzer fuzz_corpus -seed=1 -max_total_time=60
2020-06-01T19:24:21.2161016Z ==21402==Shadow memory range interleaves with an existing memory mapping. ASan cannot proceed correctly. ABORTING.
2020-06-01T19:24:21.2162128Z ==21402==ASan shadow was supposed to be located in the [0x00007fff7000-0x10007fff7fff] range.
2020-06-01T19:24:21.2162797Z ==21402==This might be related to ELF_ET_DYN_BASE change in Linux 4.12.
2020-06-01T19:24:21.2163446Z ==21402==See https://github.com/google/sanitizers/issues/856 for possible workarounds.
2020-06-01T19:24:21.2164032Z ==21402==Process memory map follows:
```

**Comment by lizan** [comment]
@asraa not likely to be related to the ELF_ET_DYN_BASE, the kernel is `5.3.0-1017-aws #18~18.04.1-Ubuntu SMP Wed Apr 8 15:12:16 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

**Comment by asraa** [comment]
Evaluation: not an kernel issue, the issue is very large non-PIE binary.

**Comment by lizan** [comment]
@asraa PIE is enabled by default, is it because too large PIE binary?

**Comment by asraa** [comment]
That's right -- I removed a handful of unused deps from the largest binaries (and now I'm working on figuring out what to do about crashes)

**Comment by asraa** [comment]
Running in plain-fuzzer mode (no asan for coverage purposes) hits the CI failure above. I think it's this https://github.com/google/oss-fuzz/issues/2328

**Comment by asraa** [comment]
Really what I want to do is to build libfuzzer without it's private libc++ library to fix the exception crash, but I'm not 100% sure how to do this. https://github.com/llvm/llvm-project/blob/6f71678ecd293c35435f7f3f313bdaf337ed798a/compiler-rt/cmake/Modules/CustomLibcxx/CMakeLists.txt#L7

Using stdlibc++ worked for most targets, but completely failed for the config_impl_test we use to generate the corpus

**Comment by asraa** [comment]
Fuzz CI passes when I build with size optimization (ASSERTs are still caught)

**Comment by asraa** [comment]
> CI failure is interesting. An easy solution is to run the gtest mode of the fuzz target when running `bazel coverage`.
> 
> I pushed the server corpus entry that triggers the `ApiBoostRetryException` and failure to the PR. It crashes with an uncaught exception. Reproduce with:
> `bazel run --config=libc++ --test_output=all --test_arg=/usr/local/google/home/asraa/git/envoy2/test/server/server_corpus --test_arg=-runs=0 //test/server/config_validation:config_fuzz_test_with_libfuzzer`
> (Replace with your own srcdir). This runs the test against the seed corpus with the fuzzing driver.
> 
> * It is important to run the target `_with_libfuzzer` that links the fuzzing main. It does not repro with using standard gtest main.
> * It must use `--config=libc++`. Without it, it succeeds.
> * If you add `--config clang-asan` in addition to `--config=libc++`, it will succeed.
> * There are some (DO NOT MERGE) `std::cerr` logs I added so that they can be seen in fuzzing mode. It indeed is getting stuck in applying the lambda.
> 
> @yanavlasov Does this indicate that this might be a compiler bug?
> stack trace shows the exception is uncaught in `tryWithApiBoosting` executing the lambda [here](https://github.com/envoyproxy/envoy/blob/49efb9841a58ebdc43a666f55c445911c8e4181c/source/common/protobuf/utility.cc#L154):
> 
> ```
> terminating with uncaught exception of type Envoy::(anonymous namespace)::ApiBoostRetryException
> ==36646== ERROR: libFuzzer: deadly signal
>     #0 0x2e75450 in __sanitizer_print_stack_trace (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x2e75450)
>     #1 0x2e20818 in fuzzer::PrintStackTrace() (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x2e20818)
>     #2 0x2e05f83 in fuzzer::Fuzzer::CrashCallback() (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x2e05f83)
>     #3 0x7f8b9648f38f  (/lib/x86_64-linux-gnu/libpthread.so.0+0x1138f)
>     #4 0x7f8b960e9427 in raise (/lib/x86_64-linux-gnu/libc.so.6+0x35427)
>     #5 0x7f8b960eb029 in abort (/lib/x86_64-linux-gnu/libc.so.6+0x37029)
>     #6 0x8aafb0a in abort_message (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x8aafb0a)
>     #7 0x8aafa35 in demangling_terminate_handler() (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x8aafa35)
>     #8 0x8aaf892 in std::__terminate(void (*)()) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x8aaf892)
>     #9 0x8aaeee5 in __cxxabiv1::failed_throw(__cxxabiv1::__cxa_exception*) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x8aaeee5)
>     #10 0x8aaee7e in __cxa_throw (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x8aaee7e)
>     #11 0x72d19a6 in Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1::operator()(google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion) const (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72d19a6)
>     #12 0x72e25a9 in decltype(std::__1::forward<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1&>(fp)(std::__1::forward<google::protobuf::Message&>(fp0), std::__1::forward<Envoy::(anonymous namespace)::MessageVersion>(fp0))) std::__1::__invoke<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1&, google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion>(Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1&, google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion&&) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72e25a9)
>     #13 0x72e24fa in void std::__1::__invoke_void_return_wrapper<void>::__call<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1&, google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion>(Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1&, google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion&&) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72e24fa)
>     #14 0x72e248a in std::__1::__function::__alloc_func<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1, std::__1::allocator<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1>, void (google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion)>::operator()(google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion&&) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72e248a)
>     #15 0x72e0e6c in std::__1::__function::__func<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1, std::__1::allocator<Envoy::MessageUtil::loadFromFile(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, google::protobuf::Message&, Envoy::ProtobufMessage::ValidationVisitor&, Envoy::Api::Api&, bool)::$_1>, void (google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion)>::operator()(google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion&&) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72e0e6c)
>     #16 0x72d4d02 in std::__1::__function::__value_func<void (google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion)>::operator()(google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion&&) const (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72d4d02)
>     #17 0x72d4c67 in std::__1::function<void (google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion)>::operator()(google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion) const (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72d4c67)
>     #18 0x72d00bb in Envoy::(anonymous namespace)::tryWithApiBoosting(std::__1::function<void (google::protobuf::Message&, Envoy::(anonymous namespace)::MessageVersion)>, google::protobuf::Message&) (/build/tmp/_bazel_bazel/b570b5ccd0454dc9af9f65ab1833764d/execroot/envoy/bazel-out/k8-fastbuild/bin/test/server/config_validation/config_fuzz_test_with_libfuzzer+0x72d00bb)
> ```
> 
> Note: There are a lot of lambas in this stack... `loadBootstrapConfig` passes a lambda to `loadBootstrap` that executes `loadFromFile`. Then `loadFromFile` passes a lambda that throws to `tryWithApiBoostring` that encapsulates the lambda with a try/catch.

**Comment by asraa** [comment]
The close was by accident. It turns out interns are also running into this issue locally https://github.com/envoyproxy/envoy/issues/11802

I got coverage without asan working, so i'll make that fix a separate PR and reopen this when when it's ready.

File: test/run_envoy_bazel_coverage.sh [hunk 4]
```diff
@@ -20,10 +20,14 @@ elif [[ -n "${COVERAGE_TARGET}" ]]; then
   COVERAGE_TARGETS=${COVERAGE_TARGET}
 else
   # For fuzz builds, this overrides to just fuzz targets.
-  COVERAGE_TARGETS=//test/... && [[ ${FUZZ_COVERAGE} == "true" ]] &&
-    COVERAGE_TARGETS="$(bazel query 'attr("tags", "fuzz_target", //test/...)')"
+  COVERAGE_TARGETS=//test/...
 fi
 
+if [[ "${FUZZ_COVERAGE}" == "true" ]]; then
```
Comment (sqkerner) [inline test/run_envoy_bazel_coverage.sh:26]: This if block is the same as the next if block.  Can they be combined?

Comment (sqkerner) [inline test/run_envoy_bazel_coverage.sh:26]: This if block has the same condition as the next one.  Can they be merged?

Comment (asraa) [reply test/run_envoy_bazel_coverage.sh:25]: Good point. Done. They were two difference "steps", so logically seperated
(filter targets, add build options)
Added a comment explaining.

**Comment by asraa** [comment]
Finally, success.

using fuzzer-no-link in bazelrc cli breaks the gtest that's used to build the corpus, hence why it's removed.

File: ci/do_ci.sh [hunk 5]
```diff
@@ -269,6 +269,18 @@ elif [[ "$CI_TARGET" == "bazel.coverage" ]]; then
...
   exit 0
+elif [[ "$CI_TARGET" == "bazel.fuzz_coverage" ]]; then
+  setup_clang_toolchain
+  echo "bazel coverage build with fuzz tests ${COVERAGE_TEST_TARGETS}"
+
+  # Reduce the amount of memory Bazel tries to use to prevent it from launching too many subprocesses.
+  # This should prevent the system from running out of memory and killing tasks. See discussion on
+  # https://github.com/envoyproxy/envoy/pull/5611.
+  [ -z "$CIRCLECI" ] || export BAZEL_BUILD_OPTIONS="${BAZEL_BUILD_OPTIONS} --local_ram_resources=12288"
+
```
Comment (lizan) [inline ci/do_ci.sh:280]: remove

File: .azure-pipelines/pipelines.yml [hunk 6]
```diff
@@ -90,21 +90,28 @@ jobs:
           ciTarget: $(CI_TARGET)
 
   - job: coverage
-    displayName: "Linux-x64 coverage"
+    displayName: "Linux-x64"
     dependsOn: ["format"]
```
Comment (lizan) [inline .azure-pipelines/pipelines.yml:94]: can you make this depends on release?

File: ci/do_ci.sh (hunk 5 above)
Comment (asraa) [reply ci/do_ci.sh:280]: done

File: .azure-pipelines/pipelines.yml (hunk 6 above)
Comment (asraa) [reply .azure-pipelines/pipelines.yml]: done

**Pull request merged**